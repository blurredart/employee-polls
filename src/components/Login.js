import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link, Outlet } from "react-router-dom";
import { doLogin } from "../actions/auth";

export default function Login()
{
    const user = useSelector(store => store.auth.id);
    return (user !== "" ?
        (<>
            <Link to='/'>Home</Link>{" "}
            <Link to='/leaderboard'>Leaderboard</Link>{" "}
            <Link to='/add'>Add Question</Link>
            <Outlet />
        </>) : <UserSelector />)
}

function UserSelector()
{

    const [user, setUser] = useState("");
    const users = useSelector(store => store.users.users);
    const status = useSelector(store => store.users.status);
    const dispatch = useDispatch();

    useEffect(() =>
    {
        if (status !== "init" && users !== {})
            setUser(Object.keys(users)[0])
    }, [status, users])

    const handleLogin = () =>
    {
        dispatch(doLogin(user));
    }
    return (status === "init" ?
        <div>Loading...</div> :
        (<>
            < select name="users" id="user-select" onChange={(e) => setUser(e.target.value)}>
                {
                    Object.keys(users).map(id =>
                    {
                        let tuser = users[id];
                        return <option value={tuser.id} key={id}>{tuser.name}</option>
                    })
                }
            </select >
            <br></br>
            <button onClick={handleLogin}>Login</button>
        </>
        ))
}