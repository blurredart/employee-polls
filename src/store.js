import { configureStore } from '@reduxjs/toolkit';
import authReducer from './reducers/authReducer.js';
import { questionsReducer } from './reducers/questionsReducer.js';
import { usersReducer } from './reducers/usersReducer.js';

export const store = configureStore({
  reducer: {
    auth: authReducer,
    users: usersReducer,
    questions: questionsReducer
  }
});

export default store