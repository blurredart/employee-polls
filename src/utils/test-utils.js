import React from 'react'
import { render } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import { MemoryRouter as Router } from 'react-router-dom'
import authReducer from '../reducers/authReducer'
import { questionsReducer } from '../reducers/questionsReducer'
import { usersReducer } from '../reducers/usersReducer'

export function renderWithProviders(
    ui,
    {
        intialEntries = ["/"],
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: {
                auth: authReducer,
                users: usersReducer,
                questions: questionsReducer
            }, preloadedState
        }),
        ...renderOptions
    } = {}
)
{
    function Wrapper({ children })
    {
        return <Provider store={store}>
            <Router initialEntries={intialEntries}>
                {children}
            </Router>
        </Provider>
    }
    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}