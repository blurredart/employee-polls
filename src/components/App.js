import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { Route, Routes } from "react-router-dom";
import notfound from '../images/404.svg';
import { fetchQuestionsAsync } from "../reducers/questionsReducer.js";
import { fetchUsersAsync } from "../reducers/usersReducer";
import AddQuestion from "./AddQuestion.js";
import Leaderboard from "./Leaderboard.js";
import Login from './Login.js';
import Question from './Question';
import Questions from "./Questions.js";

function App()
{
    const dispatch = useDispatch();

    useEffect(() =>
    {
        dispatch(fetchQuestionsAsync());
        dispatch(fetchUsersAsync());
    }, [dispatch]);

    return (
        <>

            <Routes>
                <Route element={<Login />}>
                    <Route path="/" element={<Questions />} />
                    <Route path="/questions/:id" element={<Question />} />
                    <Route path="/add" element={<AddQuestion />}></Route>
                    <Route path="/leaderboard" element={<Leaderboard />}></Route>
                </Route>
                <Route
                    path="/*"
                    element={<img src={notfound} alt="404 Page not found" />}
                />
            </Routes>
        </>);
}
export default App;
