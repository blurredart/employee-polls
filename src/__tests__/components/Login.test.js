import Login from '../../components/Login.js'
import { cleanup, screen } from '@testing-library/react';
import { Provider } from "react-redux";
import { store } from '../../store'
import { MemoryRouter as Router } from "react-router-dom";
import App from '../../components/App.js';
import { renderWithProviders } from '../../utils/test-utils.js';
import '@testing-library/jest-dom'

describe('Login page is shown for our app for supported URLs', () =>
{
    afterEach(() =>
    {
        cleanup();
    })
    it('tests at URL /', async () =>
    {
        let render = renderWithProviders(<App />, { intialEntries: ["/"] });
        expect(await screen.findByText(/Login/i)).toBeInTheDocument()
        expect(render).toMatchSnapshot();
    })

    it('tests at URL /add', async () =>
    {
        renderWithProviders(<App />, { intialEntries: ["/add"] });
        expect(await screen.findByText(/Login/i)).toBeInTheDocument()
    })

    it('tests at URL /leaderbaord', async () =>
    {
        renderWithProviders(<App />, { intialEntries: ["/leaderboard"] });
        expect(await screen.findByText(/Login/i)).toBeInTheDocument()
    })
})

describe('404 page is shown for unsupported URLs', () =>
{
    afterEach(() =>
    {
        cleanup();
    })

    it('tests at URL /signin', async () =>
    {
        renderWithProviders(<App />, { intialEntries: ["/signin"] });
        expect(await screen.findByRole('img')).toBeInTheDocument()
    })

    it('tests at URL /404', async () =>
    {
        renderWithProviders(<App />, { intialEntries: ["/404"] });
        expect(await screen.findByRole('img')).toBeInTheDocument()
    })
})