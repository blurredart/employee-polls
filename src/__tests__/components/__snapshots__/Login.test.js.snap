// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page is shown for our app for supported URLs tests at URL / 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="form-select"
        id="user-select"
        name="users"
        variant="dark"
      >
        <option
          value="sarahedo"
        >
          Sarah Edo
        </option>
        <option
          value="tylermcginnis"
        >
          Tyler McGinnis
        </option>
        <option
          value="mtsamis"
        >
          Mike Tsamis
        </option>
        <option
          value="zoshikanlu"
        >
          Zenobia Oshikanlu
        </option>
      </select>
      <button
        class="form-control btn btn-dark"
        type="submit"
      >
        Login
      </button>
    </div>
  </body>,
  "container": <div>
    <select
      class="form-select"
      id="user-select"
      name="users"
      variant="dark"
    >
      <option
        value="sarahedo"
      >
        Sarah Edo
      </option>
      <option
        value="tylermcginnis"
      >
        Tyler McGinnis
      </option>
      <option
        value="mtsamis"
      >
        Mike Tsamis
      </option>
      <option
        value="zoshikanlu"
      >
        Zenobia Oshikanlu
      </option>
    </select>
    <button
      class="form-control btn btn-dark"
      type="submit"
    >
      Login
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
